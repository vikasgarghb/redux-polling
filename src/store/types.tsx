import { JobState } from '../actions/job.types';

export interface AppState {
  job: JobState;
}
