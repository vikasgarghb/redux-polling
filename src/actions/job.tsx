import { ON_CREATE_JOB, ON_GET_STATUS, ON_COMPLETE_PROCESSING } from './job.types';

export function createJob() {
  return {
    type: ON_CREATE_JOB,
  };
}

export function getJobStatus(jobId: string) {
  return {
    type: ON_GET_STATUS,
    jobId,
  };
}

export function jobIsComplete() {
  return {
    type: ON_COMPLETE_PROCESSING,
  };
}
