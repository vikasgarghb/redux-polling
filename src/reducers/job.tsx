import {
  JobState,
  JobStatus,
  ON_ERROR,
  ON_SUCCESS,
  ON_CREATE_JOB,
  ON_SUCCESS_GET_STATUS,
  ON_GET_STATUS,
  ON_COMPLETE_PROCESSING,
} from '../actions/job.types';

export const defaultJob: JobState = {
  fetching: false,
  processing: false,
};

const mapJobStatus = (state: JobState, jobStatus: JobStatus) => ({
  job: {
    ...state.job,
    status: jobStatus,
  },
});

/**
 * Reducer
 */
export default function jobReducer(state: JobState = defaultJob, action: any) {
  switch (action.type) {
    case ON_ERROR:
      return { ...state, ...mapJobStatus(state, JobStatus.Failed), processing: false };
    case ON_SUCCESS:
      return { ...state, job: action.payload };
    case ON_CREATE_JOB:
      return { ...state, processing: true };
    case ON_SUCCESS_GET_STATUS:
      return { ...state, ...mapJobStatus(state, action.payload), fetching: false };
    case ON_GET_STATUS:
      return { ...state, fetching: true };
    case ON_COMPLETE_PROCESSING:
      return { ...state, processing: false };
    default:
      return state;
  }
}
