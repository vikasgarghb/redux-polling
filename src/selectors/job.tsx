import { AppState } from '../store/types';
import { get } from 'lodash';

export const getJob = (state: AppState) => get(state, 'job.job', {});
