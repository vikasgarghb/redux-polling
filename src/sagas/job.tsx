import {
  ON_SUCCESS,
  ON_ERROR,
  ON_SUCCESS_GET_STATUS,
  ON_CREATE_JOB,
  ON_GET_STATUS,
  JobStatus,
} from '../actions/job.types';
import { select, put, delay, takeLatest, all } from 'redux-saga/effects';
import { getJob } from '../selectors/job';
import { post, get } from './common';
import { getJobStatus, jobIsComplete } from '../actions/job';

export function* createJob(action) {
  const { payload } = action;

  yield post({
    url: '/job',
    onSuccess: ON_SUCCESS,
    onError: ON_ERROR,
    data: payload,
  });
}

export function* startPollingForStatus() {
  const job = yield select(state => getJob(state));
  yield put(getJobStatus(job.id));
}

export function* getStatus(action) {
  const { jobId } = action;

  yield get({
    url: `/job/${jobId}/status`,
    onSuccess: ON_SUCCESS_GET_STATUS,
    onError: ON_ERROR,
  });
}

export function* isJobComplete() {
  const job = yield select(state => getJob(state));
  if (job.status === JobStatus.Completed) {
    return yield put(jobIsComplete());
  }

  if ([JobStatus.Created, JobStatus.Processing].includes(job.status)) {
    yield delay(2000);
    return yield put(getJobStatus(job.id));
  }

  yield put({
    type: ON_ERROR,
  });
}

export default function* root() {
  return yield all([
    takeLatest(ON_CREATE_JOB, createJob),
    takeLatest(ON_GET_STATUS, getStatus),
    takeLatest(ON_SUCCESS, startPollingForStatus),
    takeLatest(ON_SUCCESS_GET_STATUS, isJobComplete),
  ]);
}
