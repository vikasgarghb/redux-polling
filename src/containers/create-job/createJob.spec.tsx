import React from 'react';
import { shallow } from 'enzyme';

import { CreateJob, IProps } from '.';
import { JobStatus } from '../../actions/job.types';

describe('CreateJob', () => {
  const props: IProps = {
    handleCreateJob: jest.fn(),
  };

  afterEach(() => {
    props.job = undefined;
  });

  it('Should match snapshot', () => {
    const wrapper = shallow(<CreateJob {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot with processing icon', () => {
    props.job = { id: '123', status: JobStatus.Processing };
    const wrapper = shallow(<CreateJob {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot with job completed', () => {
    props.job = { id: '123', status: JobStatus.Completed };
    const wrapper = shallow(<CreateJob {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('Functions', () => {
    it('should call handleCreateJob', () => {
      const wrapper = shallow(<CreateJob {...props} />);
      const instance = wrapper.instance() as CreateJob;
      instance.handleCreateButton();
      expect(props.handleCreateJob).toHaveBeenCalled();
    });
  });
});
