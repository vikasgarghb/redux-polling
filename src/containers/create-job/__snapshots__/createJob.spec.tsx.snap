// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateJob Should match snapshot 1`] = `
<styled.div>
  <button
    onClick={[Function]}
  >
    Create Job
  </button>
</styled.div>
`;

exports[`CreateJob should match snapshot with job completed 1`] = `
<styled.div>
  <button
    onClick={[Function]}
  >
    Create Job
  </button>
  <span>
    Job Created!!!
  </span>
</styled.div>
`;

exports[`CreateJob should match snapshot with processing icon 1`] = `
<styled.div>
  <button
    onClick={[Function]}
  >
    Create Job
  </button>
  <Processing
    title="Processing"
  />
</styled.div>
`;
